Integrate with Nerdbank.GitVersioning

scott-xu opened the pull request.

scott-xu committed:
GeneratePackageOnBuild
IncludeSymbols

scott-xu committed:
Add packages to artifacts

scott-xu committed:
Merge branch 'develop' into generate-nuget-package

scott-xu committed:
Merge branch 'develop' into generate-nuget-package

scott-xu committed:
Update src/Renci.SshNet/Renci.SshNet.csproj

Co-authored-by: Rob Hague <rob.hague00@gmail.com>

scott-xu committed:
Merge branch 'develop' into generate-nuget-package

scott-xu committed:
Merge branch 'develop' into generate-nuget-package

scott-xu committed:
Update Renci.SshNet.csproj

Co-authored-by: Rob Hague <rob.hague00@gmail.com>

scott-xu committed:
Merge branch 'develop' into generate-nuget-package

scott-xu committed:
Merge branch 'develop' into generate-nuget-package

scott-xu committed:
Merge branch 'develop' into generate-nuget-package

scott-xu committed:
Delete build/nuget/SSH.NET.nuspec

scott-xu committed:
Delete SSH.NET.nuspec from .sln file

scott-xu committed:
Merge branch 'develop' into generate-nuget-package

scott-xu committed:
Update Renci.SshNet.csproj

scott-xu committed:
Update build.proj

scott-xu committed:
init nbgv

scott-xu committed:
Merge branch 'generate-nuget-package' into nbgv

scott-xu committed:
Include version.json in "Solution Items"

scott-xu committed:
Merge branch 'sshnet:develop' into nbgv

scott-xu committed:
Update Directory.Build.props

scott-xu committed:
define publicReleaseRefSpec

scott-xu committed:
update version.json

scott-xu committed:
update version.json

scott-xu committed:
set cloud build number

scott-xu committed:
fix https://github.com/sshnet/SSH.NET/issues/1292

scott-xu committed:
remove unexpected code format

Rob-Hague commented:
Interesting. Would this change how one creates a release? If so, what would be the steps required?

scott-xu commented:
From reading the doc about nbgv, the first step is to run `nbgv prepare-release`. See help below
```
C:\Users\scott\source\repos\scott-xu\SSH.NET> nbgv prepare-release --help
Description:
  Prepares a release by creating a release branch for the current version and adjusting the version on the current branch.

Usage:
  nbgv prepare-release [<tag>] [options]

Arguments:
  <tag>  The prerelease tag to apply on the release branch (if any). If not specified, any existing prerelease tag will be removed. The preceding hyphen may be omitted.

Options:
  -p, --project <project>                The path to the project or project directory. The default is the current directory.
  --nextVersion <nextVersion>            The version to set for the current branch. If omitted, the next version is determined automatically by incrementing the current version.
  --versionIncrement <versionIncrement>  Overrides the 'versionIncrement' setting set in version.json for determining the next version of the current branch.
  -f, --format <json|text>               The format to write information about the release. Allowed values are: text, json. The default is text.
  -?, -h, --help                         Show help and usage information
```
Then grab a cup of coffee and wait the Appveyor to build the project, generate the nuget package, test the project, and finally [deploy the nuget package](https://www.appveyor.com/docs/deployment/nuget/) to nuget.org.
Obviously we need to config the Appveyor properly first.

scott-xu committed:
Delete version from csproj

scott-xu committed:
move nbgv from Directory.Build.props to Renci.SshNet.csproj as only this particular project needs versioning.
include package version in ThisAssembly and use nuget package version for the SSH client version. Since we define the precision of nuget package to "build", it has 3 digits. Nuget package version is unique which should be suffient.

Rob-Hague commented:
I get a weird error when building this branch

```
1>C:\Users\rhague\.nuget\packages\nerdbank.gitversioning\3.7.48-alpha\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018: The "Nerdbank.GitVersioning.Tasks.GetBuildVersion" task failed unexpectedly.
1>C:\Users\rhague\.nuget\packages\nerdbank.gitversioning\3.7.48-alpha\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018: Nerdbank.GitVersioning.GitException: Got a com instead of a tag when opening object 1c7166a002d7633fe1595b3df4634ba0ef6e3138
1>C:\Users\rhague\.nuget\packages\nerdbank.gitversioning\3.7.48-alpha\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018:    at Nerdbank.GitVersioning.ManagedGit.GitRepository.TryGetObjectByPath(GitObjectId sha, String objectType, Stream& value)
1>C:\Users\rhague\.nuget\packages\nerdbank.gitversioning\3.7.48-alpha\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018:    at Nerdbank.GitVersioning.ManagedGit.GitRepository.TryGetObjectBySha(GitObjectId sha, String objectType, Stream& value)
1>C:\Users\rhague\.nuget\packages\nerdbank.gitversioning\3.7.48-alpha\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018:    at Nerdbank.GitVersioning.ManagedGit.GitRepository.<LookupTags>g__HandleCandidate|44_0(GitObjectId pointsAt, String tagName, Boolean isPeeled, <>c__DisplayClass44_0& )
1>C:\Users\rhague\.nuget\packages\nerdbank.gitversioning\3.7.48-alpha\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018:    at Nerdbank.GitVersioning.ManagedGit.GitRepository.LookupTags(GitObjectId objectId)
1>C:\Users\rhague\.nuget\packages\nerdbank.gitversioning\3.7.48-alpha\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018:    at Nerdbank.GitVersioning.Managed.ManagedGitContext.get_HeadTags()
1>C:\Users\rhague\.nuget\packages\nerdbank.gitversioning\3.7.48-alpha\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018:    at Nerdbank.GitVersioning.VersionOracle..ctor(GitContext context, ICloudBuild cloudBuild, Nullable`1 overrideVersionHeightOffset)
1>C:\Users\rhague\.nuget\packages\nerdbank.gitversioning\3.7.48-alpha\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018:    at Nerdbank.GitVersioning.Tasks.GetBuildVersion.ExecuteInner() in D:\a\1\s\src\Nerdbank.GitVersioning.Tasks\GetBuildVersion.cs:line 240
1>C:\Users\rhague\.nuget\packages\nerdbank.gitversioning\3.7.48-alpha\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
1>C:\Users\rhague\.nuget\packages\nerdbank.gitversioning\3.7.48-alpha\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext()
```

Rob-Hague commented:
> I get a weird error when building this branch

It seems to be a bug in the Nerdbank package, where it does not correctly handle lightweight tags in .git/ref/tags (e.g. 2023.0.1).

In a fresh clone, the tags are in .git/packed-refs which are handled correctly, hence it passing in CI. In my local copy of the repo, the tags are in .git/ref/tags and it is assuming they are all annotated tags.

Rob-Hague committed:
Use package from CI feed

scott-xu committed:
Merge branch 'develop' into nbgv

scott-xu committed:
Merge branch 'develop' into nbgv

scott-xu commented:
It updates the CI build number:
![image](https://github.com/sshnet/SSH.NET/assets/1913505/7c00a5e0-bc8d-4b90-af97-b4e937af8a9f)

Rob-Hague commented:
I've not forgotten about this PR, I just want to read up some more on the configuration

Rob-Hague committed:
Merge remote-tracking branch 'upstream/develop' into nbgv

Rob-Hague commented on version.json:
```diff
@@ -0,0 +1,25 @@
+    }
+  },
+  "assemblyVersion": {
+    "precision": "minor"
```
"minor" is already the default value for "precision", but perhaps this should be "build" in our case?

Rob-Hague commented on version.json:
```diff
@@ -0,0 +1,25 @@
+    "precision": "minor"
+  },
+  "nugetPackageVersion": {
+   "precision": "minor"
```
Should it be "build"? I am not exactly clear what this does (same with the "assemblyVersion":"precision")

Rob-Hague commented on version.json:
```diff
@@ -0,0 +1,25 @@
+  },
+  "release": {
+    "tagName" : "{version}",
+    "branchName" : "master",
```
I think this should be e.g.

```suggestion
    "branchName" : "release/{version}",
```

scott-xu replied:
I think the curernt release is based on `master` branch with tag?

Rob-Hague reviewed (commented):
I noticed that 2024.0.0 already has the commit info tagged to the dll (probably because of sourcelink?):

![image](https://github.com/sshnet/SSH.NET/assets/5132141/3c296963-f76a-40b5-b5d4-83bb87032667)

In that case, I am not sure if there still benefits to this change. I am not sure the release process needs to change (it seems simple enough), but as I don't make the releases I think I will have to defer to @WojciechNagorski as to whether to take this change.

scott-xu committed:
Bump version to 2024.0.1; Update version precision

scott-xu commented:
Not very familar with current release process, but it seems to be some manual work?

WojciechNagorski commented:
Yes it is manual

scott-xu commented:
I guess the current release process is based on https://github.com/sshnet/BuildTools and https://github.com/sshnet/SSH.NET/blob/develop/build/build.proj, am I right? @WojciechNagorski 
No matter merge this PR or not, I recommend delegating the release process to appveyor

WojciechNagorski commented:
Currently it is just dotnet pack right now.

Rob-Hague committed:
Merge remote-tracking branch 'upstream/develop' into nbgv

Rob-Hague committed:
Some tweaks:

- Remove the "release" section. I don't think we will use that for now
- Remove the "cloudBuild" section. Changing the CI build number doesn't seem that useful
- In the "nugetPackageVersion" section:
  - Remove "precision". It defaults to build
  - Add "semVer"=2. This makes the package version e.g. 2024.1.1-prerelease.1 instead
    of 2024.1.1-prerelease-0001
- In "assemblyVersion" change "precision" to revision. Doesn't seem to change anything, I
  was just copying nbgv's setup: https://github.com/dotnet/Nerdbank.GitVersioning/blob/main/version.json
- Make sure there are no '-' in the softwareversion string (change the test to a regex)

Rob-Hague commented:
I actually have some local changes for this, let me fix the merge and push them in

Rob-Hague committed:
Revert unnecessary test changes; remove unnecessary tests

Rob-Hague commented:
Example on a PR branch:

<details>

```c#
internal static partial class ThisAssembly {
    internal const string AssemblyConfiguration = "Release";
    internal const string AssemblyFileVersion = "2024.1.1";
    internal const string AssemblyInformationalVersion = "2024.1.1-prerelease.2+84ff4f6671";
    internal const string AssemblyName = "Renci.SshNet";
    internal const string AssemblyTitle = "SSH.NET";
    internal const string AssemblyVersion = "2024.1.1.0";
    internal static readonly global::System.DateTime GitCommitDate = new global::System.DateTime(638576102900000000L, global::System.DateTimeKind.Utc);
    internal const string GitCommitId = "84ff4f66711638341a0d856299f7a62ae4bbf38b";
    internal const bool IsPrerelease = true;
    internal const bool IsPublicRelease = false;
    internal const string NuGetPackageVersion = "2024.1.1-prerelease.2.g84ff4f6671";
    internal const string PublicKey = "0024000004800000940000000602000000240000525341310004000001000100f9194e1eb66b7e2575aaee115ee1d27bc100920e7150e43992d6f668f9737de8b9c7ae892b62b8a36dd1d57929ff1541665d101dc476d6e02390846efae7e5186eec409710fdb596e3f83740afef0d4443055937649bc5a773175b61c57615dac0f0fd10f52b52fedf76c17474cc567b3f7a79de95dde842509fb39aaf69c6c2";
    internal const string PublicKeyToken = "1cee9f8bde3db106";
    internal const string RootNamespace = "Renci.SshNet";
}
```

![image](https://github.com/user-attachments/assets/a5caa60d-f2f9-488f-922e-829e641923da)

</details>

Example on a simulated "public" branch e.g. `develop`:

<details>

```c#
internal static partial class ThisAssembly {
    internal const string AssemblyConfiguration = "Release";
    internal const string AssemblyFileVersion = "2024.1.1";
    internal const string AssemblyInformationalVersion = "2024.1.1-prerelease.3+57319f736f";
    internal const string AssemblyName = "Renci.SshNet";
    internal const string AssemblyTitle = "SSH.NET";
    internal const string AssemblyVersion = "2024.1.1.0";
    internal static readonly global::System.DateTime GitCommitDate = new global::System.DateTime(638576099590000000L, global::System.DateTimeKind.Utc);
    internal const string GitCommitId = "57319f736f1d0dbf25455f121669223d8fc48893";
    internal const bool IsPrerelease = true;
    internal const bool IsPublicRelease = true;
    internal const string NuGetPackageVersion = "2024.1.1-prerelease.3";
    internal const string PublicKey = "0024000004800000940000000602000000240000525341310004000001000100f9194e1eb66b7e2575aaee115ee1d27bc100920e7150e43992d6f668f9737de8b9c7ae892b62b8a36dd1d57929ff1541665d101dc476d6e02390846efae7e5186eec409710fdb596e3f83740afef0d4443055937649bc5a773175b61c57615dac0f0fd10f52b52fedf76c17474cc567b3f7a79de95dde842509fb39aaf69c6c2";
    internal const string PublicKeyToken = "1cee9f8bde3db106";
    internal const string RootNamespace = "Renci.SshNet";
}
```

![image](https://github.com/user-attachments/assets/1ee33d42-3fed-4cbc-81f5-ec9aa9951482)

</details>

Rob-Hague commented:
The appveyor nuget feed is in operation:

![image](https://github.com/user-attachments/assets/8f6e3095-4c4f-4bb1-9e62-c433d5a40159)
